picoquic/picoquic.h

```diff
@@ -190,6 +192,22 @@ typedef void (*picoquic_stream_data_cb_fn)(picoquic_cnx_t* cnx,
 typedef void (*cnx_id_cb_fn)(picoquic_connection_id_t cnx_id_local,
     picoquic_connection_id_t cnx_id_remote, void* cnx_id_cb_data, picoquic_connection_id_t * cnx_id_returned);
 
+/* Will be called to verify that the given data corresponds to the given signature.
+ * This callback and the `verify_ctx` will be set by the `verify_certificate_cb_fn`.
+ * If `data` and `sign` are empty buffers, an error occurred and `verify_ctx` should be freed.
+ * Expect `0` as return value, when the data matches the signature.
+ */
+typedef int (*verify_sign_cb_fn)(void* verify_ctx, ptls_iovec_t data, ptls_iovec_t sign);
+/* Will be called to verify a certificate of a connection.
+ * The arguments `verify_sign` and `verify_sign_ctx` are expected to be set, when the function returns `0`.
+ * See `verify_sign_cb_fn` for more information about these arguments.
+ */
+typedef int (*verify_certificate_cb_fn)(void* ctx, picoquic_cnx_t* cnx, ptls_iovec_t* certs, size_t num_certs,
+                                        verify_sign_cb_fn* verify_sign, void** verify_sign_ctx);
+
+/* Is called to free the verify certificate ctx */
+typedef void (*free_verify_certificate_ctx)(void* ctx);
+
```

Comment:
Maybe the names verify_sign_cb_fn, verify_certificate_cb_fn, or free_verify_certificate_ctx  should have a "picoquic_" prefix to avoid collision with other packages.

Do we actually need a free_verify_certificate_ctx call back? What is the advantage of managing the memory from within the stack, versus letting the application deal with it?

Reply:
I can do the prefixing.

Hmm, in my opinion it is better to specify a free function that is called when picoquic is freed. Otherwise, I would need to store the ctx a second time in the application layer and would need to free this, after picoquic was freed. Do you understand my point here?

Reply:
OK. But we don't do that anywhere else, so there is a small issue of consistency. Maybe I should be adopting the same approach as Kazuho. Instead of specifying the cb function directly in picotls, he asks you to provide a 
~~~
    struct { function_type cb } function_cb;
~~~
Then the application can allocate a larger struct, and use offset_off or container_off to get the corresponding data. I don't like it too much, but it does make the stack a bit simpler.

Reply:
Just to be clear, your code is fine for now. If we change something, it will have to be done globally.

Reply:
Hmm. But that would not solve the problem with deallocating the context. With picotls that works, because the structs are stored in the global context and you can access them from the outside.

We could define certain values for the callback which mean that the context should be freed? Like picotls is doing it for the verify_sign callback.